enhancements/update/upgrade-without-registry.md

```diff
@@ -0,0 +1,893 @@
+including the corresponding `additionalimagestores` field, and then reload it
+using the equivalent of `systemctl reload crio.service`.
+
+The changes to pin the images will be done in a
+`/etc/crio/crio.conf.d/pin-upgrade.conf` file, something like this:
+
+```toml
+pinned_images=[
+  "quay.io/openshift-release-dev/ocp-release@sha256:...",
+  "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...",
+  "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...",
+  ...
+]
+```
+
+These files aren't tracked by the machine config operator, but the CRI-O service
+needs to be reloaded when they are created or changed. To support that a new
+filed will be added to the `ContainerRuntimeConfig` object, for example:
+
+```yaml
+apiVersion: machineconfiguration.openshift.io/v1
+kind: ContainerRuntimeConfig
+metadata:
+  name: pin-upgrade
+spec:
+  containerRuntimeConfig:
+    pinnedImages:
+    - quay.io/openshift-release-dev/ocp-release@sha256:...
+    - quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...
+    - quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...
+    ...
+```
+
+When the new `pinnedImages` field is added or changed the machine config
+operator will need to create or update the corresponding
+`/etc/crio/crio.conf.d/pin-upgrade.conf` file, and then reload it using the
+equivalent of `systemctl reload crio.service`.
+
+#### Tool to create the upgrade bundle
+
+The upgrade bundle will be created by an engineer in a partner's factory using a
+new `oc adm upgrade create bundle` command. This engineer will first determine
+the target version number, for example 4.13.7. Note that doing this will
+probably require access to the upgrade service available in api.openshift.com.
+Finding that upgrade version number is outside of the scope of this enhancement.
+
+The engineer will then need internet access and a Linux machine where she can
+run `oc adm upgrade create bundle`, for example:
+
+```bash
+$ oc adm upgrade create bundle \
+--arch=x86_64 \
+--version=4.13.7 \
+--pull-secret=/my/pull/secret.txt \
+--output=/my/bundle/dir
+```
+
+The `oc adm upgrade bundle` command will find the list of image references that
+make up the release, doing the equivalent of this:
+
+```bash
+$ oc adm release info \
+quay.io/openshift-release-dev/ocp-release:4.13.7-x86_64 -o json | \
+jq '.references.spec.tags[].from.name'
+```
+
+In addition to the release images the tool will also support explicitly adding
+custom images. For example:
+
+```bash
+$ oc adm upgrade create bundle \
+--arch=x86_64 \
+--version=4.13.7 \
+--pull-secret=/my/pull/secret.txt \
+--extra-image=quay.io/my-company/my-workload1 \
+--extra-image=quay.io/my-company/my-workload2 \
+...
+--output=/my/bundle/dir
+```
+
+This is intended for situations where the user wants to use the same upgrade
+mechanism for her own images.
+
+The command will then bring up a temporary image registry server, embedded into
```

Comment:
Why does it need to start a local registry instead of copying the files directly to disk?

Reply:
We explored several alternatives for this and this seems to be the best one.

A plain file copy is feasible only if the target directory is empty and nothing is using it. That is why we are suggesting to add full support to the `additionalimagestores` setting in `storage.conf`. If that was usable (via `ContainerRuntimeConfig`) then we could create such an additional directory, copy the files there and reload CRI-O. The size of the (uncompressed) bundle for this would be approximately 32 GiB, and it would be there for ever, at least till the next upgrade, because those additional directories are read only.

It is also possible to copy the images using the equivalent of `skopeo copy containers-storage:... containers-storage: ...`. That doesn't need need the permanent additional directory, and doesn't need to shutdown or reload CRI-O, but it does need those 32 GiB temporarily, while the copy is in progress.

Another possibility is to use the equivalent of `skopeo copy docker://... containers-storage:...`. That is where the temporary registry comes in place. The advantage is that the format used by the registry to store the images is more efficient: it only needs 16 GiB for a release. That reduces the size of the bundle and the space required in the node.

An improvement over that last possibility is to use the CRI-O gRPC API, the equivalent of `circtl pull ...`. It doesn't improve performance or reduces the required size, but it means that there is one less component needed (no need for `skopeo`) and reduces the risks: CRI-O will be writing the images to the disk itself, so no risk of format mismatch.

We aren't ruling out any of the above possibilities, consider them as implementation details, but we think that the last one is the better overall.

Reply:
Done.